Add documentation links for prefab samples

ggfan opened the pull request:
a couple of minor changes:

1. rename curl-ssl to prefab-using: the purpose is to show prefab
2. update README.md to include formal documentation links

PTAL.

DanAlbert commented on prefab/prefab-using/README.md:
```diff
@@ -1,26 +1,45 @@
 * Add the dependencies into the native module's [android.mk] or [CMakeLists.txt]
 
-From here on, the headers and native libs in the dependencies are [ready to use]!
+From here on, the [headers] and [native libs] in the dependencies are ready to use!
```
Just "libraries", not "native libs"

DanAlbert commented on prefab/prefab-using/README.md:
```diff
@@ -1,26 +1,45 @@
-From here on, the headers and native libs in the dependencies are [ready to use]!
+From here on, the [headers] and [native libs] in the dependencies are ready to use!
+
+If you want to understand more about C/C++ dependency AARs, simply unzip the AAR, 
+check out the `prefab` directory; and go to [C/C++ dependencies] documentation page.
```
Remove "simply" and replace the semicolon with a comma.

DanAlbert reviewed (commented):
How about `prefab-curl-ssl`, or `prefab-https`? `prefab-using` doesn't tell me much.

I'd also argue against renaming it in the first place, since it'll break a bunch of references to this sample. If it really must be renamed, we probably should leave the old directory in place with a `README.md` pointing to the new directory.

ggfan reviewed (commented):
addressed some comments, and explained some.  Changed the sample names to be:
  - using-native-dependency
  - generating-native-dependency

to make it symmetric in the names.  The purposes of the samples are to show how to use Prefab / Dependency AARs,  the curl/openSSL/JsonCpp are the just happened to be the example AARs used to demonstrate the prefab framework in AGP 4.0+.

ggfan commented on prefab/prefab-using/README.md:
```diff
@@ -1,26 +1,45 @@
+If you want to understand more about C/C++ dependency AARs, simply unzip the AAR, 
+check out the `prefab` directory; and go to [C/C++ dependencies] documentation page.
+
+The prefab-using sample uses [curl], [OpenSSL], and [JsonCpp] dependencies to display a list of the most
```
done.

DanAlbert replied:
This wasn't done

ggfan commented:
renamed the sample directories, kept the original directory as empty with a README.md there, added a top level  README.md as a sample navigation map. PTAL, thank you!

ggfan commented:
this is ready to go, kindly let me know for any concerns: like to merge it in today, thanks!

DanAlbert commented on prefab/dependency-creation/README.md:
```diff
@@ -0,0 +1,74 @@
+===============================
+
+This sample shows how to export native libraries to an AAR
+using the [Prefab] format. Generally speaking, to export native libraries in an AAR 
```
Drop "generally speaking"

DanAlbert commented on prefab/dependency-creation/README.md:
````diff
@@ -0,0 +1,74 @@
+
+* Enable the `prefabPublishing` build feature
+```
+    android.buildFeatures.prefabPublising true
````
typo

DanAlbert commented on prefab/prefab-publishing/README.md:
```diff
@@ -1,44 +1,4 @@
-# Prefab AAR Publishing
+Note:
+   This directory is deprecated and moved into
+   https://github.com/android/ndk-samples/tree/main/prefab/curl-ssl/dependency-creation.
```
This URL is wrong

DanAlbert commented on prefab/dependency-integration/settings.gradle:
```diff
@@ -0,0 +1,2 @@
+rootProject.name='prefab-using'
```
Project name is wrong

DanAlbert commented on prefab/dependency-integration/README.md:
```diff
@@ -0,0 +1,87 @@
+==============================
+
+This sample shows how to import [C/C++ dependencies] from Google Maven to your Android project.
+Those dependencies are C/C++ libraries packed with [Prefab] format into an AAR.
```
Those dependencies are C/C++ headers and packages packaged inside an AAR using the Prefab package format.

DanAlbert commented on prefab/dependency-integration/README.md:
```diff
@@ -0,0 +1,87 @@
+This sample shows how to import [C/C++ dependencies] from Google Maven to your Android project.
+Those dependencies are C/C++ libraries packed with [Prefab] format into an AAR.
+
+[Prefab] could be enabled from Android Gradle Plugin 4.0+. In roder to use the feature,
```
Prefab is supported beginning with AGP 4.0.

To use the feature, follow these three steps:

DanAlbert commented on prefab/dependency-integration/README.md:
```diff
@@ -0,0 +1,87 @@
+  done with `android.buildFeatures.prefab=true` in the application's build.gradle file.
+* Declare the dependencies in the app module's [build.gradle] in the same way you as would do with Kotlin/Java
+  dependencies.
+* Add the dependencies into the native module's [android.mk] or [CMakeLists.txt]
```
Android.mk

ggfan replied:
done

DanAlbert commented on prefab/dependency-creation/README.md:
```diff
@@ -0,0 +1,74 @@
+
+**Note** that you could use any other ways to generate the prefab AAR, as long as the AAR contains a
+legal prefab package.
+
```
Is this helpful? That's true of everything in the AAR.

ggfan replied:
this helps in the sense that "this is just one example" of how to generate prefab AAR(s), but not the ONLY way to do it.

DanAlbert replied:
But if we're not actually pointing people at other ways to do it this isn't really helpful. "You can figure out another way to do this on your own" is true of the entire AAR, but we don't call that out elsewhere because it's obvious.

DanAlbert replied:
If you want to point people at something, https://android.googlesource.com/platform/tools/ndkports/ and vcpkg are other options.

DanAlbert commented on prefab/dependency-creation/README.md:
````diff
@@ -0,0 +1,74 @@
+        headers "src/main/cpp/include"
+    }
+```
+* (Optional)If the exported native modules are NOT used by local Kotlin/Java modules in this AAR,
+  you could temporarily avoid [the unnecessary duplication inside AAR]
````
Use **strong**, not YELLING

Missing space after `)`

What does "local" mean here? I think you want "required by the Kotlin/Java modules of the AAR"? Modules? Libraries? APIs? Not sure how to describe them.

Remove "temporarily", add link to the bug.

ggfan replied:
updated: yeah, local kotlin/java modules means kotlin/java code inside this same AAR.

DanAlbert commented on prefab/dependency-creation/README.md:
````diff
@@ -0,0 +1,74 @@
+    }
+```
+This sample builds a trivial library called `mylibrary` (libmylibrary.so), and packs it into 
+mylibrary.aar. From here you could distribute the AAR via Maven center or in other ways.
````
Maven Central, not Maven center.

ggfan replied:
done

DanAlbert commented on prefab/curl-ssl/README.md:
```diff
@@ -1,68 +1,4 @@
-# curl-ssl
+Note:
+   This directory is deprecated and moved into
+   https://github.com/android/ndk-samples/tree/main/prefab/curl-ssl/dependency-integration.
```
This URL is wrong. I think this might also get formatted as a code block?

ggfan replied:
fixed, thanks!

DanAlbert reviewed (changes requested):
I'm still against the rename. Breaking existing URLs is really frustrating for users. The README helps when the existing URL points to the directory or the readme, but it doesn't help for links to anything else in the project.

ggfan commented on prefab/dependency-creation/mylibrary/build.gradle:
```diff
@@ -39,14 +37,24 @@ android {
         }
     }
+
+    // Avoid packing the unnecessary libraries into final AAR, for details
+    // refer to https://issuetracker.google.com/issues/168777344#comment5
+    packagingOptions {
+//        exclude("**/libmylibrary.so")
+//        exclude("**/libc++_shared.so")
+    }
```
this is needed, uncommented.

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,81 @@
+==============
+
+The samples here are to demontrate how to use [the Prefab feature] in Android Gradle Plugin.
+Generally speaking, Prefab is a mechanism to make C/C++ libraries available to the gradle's
+C/C++ build system, and the requirement is that the C/C++ libraries be organized in the
```
ndk-build / CMake are part of android gradle plugin build system, correct? Prefab get integrated into gradle, ndk-build/cmake get integerated into the gradle.

DanAlbert replied:
Gradle has its own entirely unrelated C++ build system.

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,81 @@
+pre-defined format and packed inside AARs. Prefab's flow in Android Gradle Plugin, at high level,
+could be viewed as:
+
+1. Use Prefab to package C/C++ libraries (including the header files), called modules, into AAR
```
changed. AGP's change to support prefab in 4.1 pack C/C++ modules into AAR.

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,81 @@
+* dependency-creation is for how to create Prefab dependency into an AAR
+* dependency-integration is for how to use Prefab AARs hosted in Google Maven Center
+
+curl-ssl and prefab-publishing directories are migrated into the samples named above.
```
Removed the final sentence.  Same as Teapots examples, need a top level README to tell what are in the sub-directories. Like to keep to help people come here first time, thank you!

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,81 @@
+[the Prefab feature]: https://developer.android.com/studio/build/native-dependencies?buildsystem=cmake&agpversion=4.0
+[the formal AAR documentations]: https://developer.android.com/studio/projects/android-library
+
+Prefab Availability
+-------------------
+
+Prefab has been added into Android Gradle Plugin since version 4:
+* Using Prefab AARs: AGP 4.0+ (with update in 4.1+)
+* Generating Prefab AARs: AGP 4.1+
```
The overall feature around prefab is the prefab: prefab hooks are added to AGP to make it work; prefab needs AAR to be in the prefab format ( for the C/C++ subdirectories ).  it is all prefab; prefab support is in AGP, Prefab, Prefab dependency, prefab AAR, native dependency AAR all bear the same meaning. User do not know/confuse with that much of details to use it. This is prefab feature, and sample is said in the beginning about "prefab". let's stick with it?

DanAlbert replied:
>  AGP, Prefab, Prefab dependency, prefab AAR, native dependency AAR all bear the same meaning

They don't. Conflating these things is going to cause a lot of confusion.

enh-google commented:
> this is ready to go, kindly let me know for any concerns: like to merge it in today, thanks!

this change seems like a bad idea? i think if we want a simple "how to add native dependencies" sample (which we might!) then this shouldn't be it. this one is really quite complicated because setting up curl and ssl certificate stores is hard. and most of the bulk of this sample is about that. so a rename that added "certificates" to the existing "curl" and "ssl" might make sense, but this is really a "bad" example of just using native dependencies because it's so full of distracting stuff that's only meaningful for the special case of curl+ssl.

would it make more sense to have a trivial example (maybe just jsoncpp) instead? _that_ could usefully just show the native dependencies side of things without all this distracting (and mostly irrelevant) detail...

ggfan commented:
@enh-google agreed that this sample to show prefab usage and ssl, special treatment causes whole lot of confusion and distraction.  Another sample would help very much to just show the usage as #749: however from that discussion, it pretty much arrived to the conclusion that curl-ssl example is the one for that.  

Now everyone agrees that another sample is the solution, let's go for that, it paints a very clear picture:
  - one sample shows how to generate a simple mylibrary
  - one sample show how to use that one from Maven Center
  - other samples to show specific libraries that need special attention/handling <-- and this category would grow along time.

enh-google commented:
> @enh-google agreed that this sample to show prefab usage and ssl, special treatment causes whole lot of confusion and distraction. Another sample would help very much to just show the usage as #749: however from that discussion, it pretty much arrived to the conclusion that curl-ssl example is the one for that.
> 
> Now everyone agrees that another sample is the solution, let's go for that, it paints a very clear picture:
> 
> * one sample shows how to generate a simple mylibrary
> * one sample show how to use that one from Maven Center
> * other samples to show specific libraries that need special attention/handling <-- and this category would grow along time.

yeah, sgtm. curl+ssl is definitely a popular combo, but it's also by far the hardest (so far, anyway!), and i don't want to scare people off...

certainly when i'm "copy and pasting my way to success"(TM) i like to have multiple small focused examples that each do one thing rather than one big example where i have to work out by trial and error which are the active ingredients and which are just implementation details :-)

DanAlbert commented:
I knew that idea sounded familiar: https://github.com/android/ndk-samples/pull/686

DanAlbert commented:
I think we can revive that to cover the simple case, leave the existing two samples as-is to avoid breaking existing URLs, and finish the docs work being done here and that leaves everyone happy?

If you want to go ahead and delete the cdep sample separately from this change, that part LGTM.

ggfan commented:
https://github.com/android/ndk-samples/pull/756 is up for that jsoncpp sample.

ggfan commented:
Hey @DanAlbert: just Integrated the prefab-dependency ( jasoncpp ) example with a rebase,  updated README.md files, ready to be reviewed!
Thank you!

DanAlbert commented on prefab/prefab-publishing/README.md:
````diff
@@ -1,27 +1,55 @@
+    }
+```
+This sample builds a trivial library called `mylibrary` (libmylibrary.so), and packs it into 
+mylibrary.aar. From here you could distribute the AAR via Maven Center or in other ways.
````
via Maven (if you want to name a specific repository for some reason, it's Maven Cent**ral**)

"or however you distribute AARs" ("or in other ways" sounds to me like we know other ways and just can't be bothered to explain them)

ggfan replied:
done

DanAlbert commented on prefab/prefab-publishing/README.md:
````diff
@@ -1,27 +1,55 @@
+```
+This sample builds a trivial library called `mylibrary` (libmylibrary.so), and packs it into 
+mylibrary.aar. From here you could distribute the AAR via Maven Center or in other ways.
+For how to use the generated dependency AAR, refer to [the prefab-dependency sample] in this repo;
````
To learn how to use the...

ggfan replied:
done

DanAlbert commented on prefab/prefab-publishing/README.md:
```diff
@@ -1,27 +1,55 @@
+This sample builds a trivial library called `mylibrary` (libmylibrary.so), and packs it into 
+mylibrary.aar. From here you could distribute the AAR via Maven Center or in other ways.
+For how to use the generated dependency AAR, refer to [the prefab-dependency sample] in this repo;
+for the official documentation about Prefab in AGP, go to [C/C++ dependencies] website.
```
see the [C/C++ dependencies] documentation.

(it's a web page, not a web site)

DanAlbert commented on prefab/prefab-publishing/README.md:
```diff
@@ -1,27 +1,55 @@
+For how to use the generated dependency AAR, refer to [the prefab-dependency sample] in this repo;
+for the official documentation about Prefab in AGP, go to [C/C++ dependencies] website.
+
+**Note** that you could use any other ways to generate the prefab AAR, as long as the AAR contains a
+legal prefab package.
```
I don't this helps anyone. That's true of everything contained in the AAR

ggfan replied:
deleted.

DanAlbert commented on prefab/prefab-publishing/mylibrary/build.gradle:
```diff
@@ -14,17 +14,15 @@
-plugins {
-    id 'com.android.library'
-}
+apply plugin: 'com.android.library'
```
Why the regression? This is the legacy plugin syntax: https://docs.gradle.org/current/userguide/plugins.html#sec:old_plugin_application

ggfan replied:
sample in this repo adopts "apply plugin".

DanAlbert replied:
That's a reason for those to be fixed, not for this one to regress.

DanAlbert commented on prefab/prefab-publishing/mylibrary/build.gradle:
```diff
@@ -39,14 +37,26 @@ android {
         }
     }
 
+    // Enable prefab intergrated inside AGP 4.1+
```
Enable generation of Prefab packages and include them in the library's AAR.

DanAlbert commented on prefab/prefab-publishing/mylibrary/build.gradle:
```diff
@@ -39,14 +37,26 @@ android {
         prefabPublishing true
     }
 
+    // Configure prefab for module headers
```
Include the "mylibrary" module from the native build system in the AAR, and export the headers in src/main/cpp/include to its consumers.

ggfan replied:
updated

DanAlbert commented on prefab/prefab-publishing/mylibrary/build.gradle:
```diff
@@ -39,14 +37,26 @@ android {
         }
     }
+
+    // Avoid packing the unnecessary libraries into final AAR, for details
```
Period, not comma

ggfan replied:
done! thx

DanAlbert commented on prefab/prefab-publishing/mylibrary/build.gradle:
```diff
@@ -39,14 +37,26 @@ android {
+
+    // Avoid packing the unnecessary libraries into final AAR, for details
+    // refer to https://issuetracker.google.com/issues/168777344#comment5
+    // Note that if your AAR's Java code uses any C/C++ code support in the same AAR,
+    //      the workaround could not be used.
```
This isn't really accurate. If the Java code uses libotherlibrary.so it's fine to exclude libmylibrary.so (and, frankly, if they're using libc++_shared in their JNI code their AAR is probably unusable: https://android.googlesource.com/platform/ndk/+/master/docs/user/middleware_vendors.md#for-java-middleware-with-jni-libraries).

How about:

Note that if your AAR also contains Java/Kotlin APIs, you should not exclude libraries that are used by those APIs.

ggfan replied:
Right now, the "exclude" is a global switch: once turned on, all *.so files will be excluded ( otherlibrary and mylibrary ), which is very dangerous. this switch is not prefab specific. but for pure native AAR purpose, there is no java code, could use this workaround.

ggfan replied:
the "exclude" is global: it will exclude any *.so files packed into AAR's "jni" folder: it is kind of dangerous, would break the AARs if those AARs do have some C++ code supporting their java code ( assume they have some Java API exported too ).  But for pure native AARs ( without any java code ), we could use this workaround.

DanAlbert replied:
Still needs to be fixed.

DanAlbert replied:
> the "exclude" is global: it will exclude any *.so files packed into AAR's "jni" folder

Not with the wildcard you've given. It only excludes the libraries that match the wildcard.

ggfan replied:
yeah, make sense now. fixed by removing the second statement.

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,80 @@
+The samples here are to demontrate how to use [the Prefab feature] in Android Gradle Plugin.
+Generally speaking, Prefab is a mechanism to make C/C++ libraries available to the gradle's
+C/C++ build system, with the requirement that the C/C++ libraries are organized in the
+[Prefab format] and packed inside AARs. Prefab's flow in Android Gradle Plugin, at high level,
+could be viewed as:
+
+1. Use AGP to pack C/C++ libraries (including the header files), called modules, into AAR
+1. Distribute AARs with maven center or other mechanisms, but this is not Prefab specific
+1. Add AAR dependency to your application, now you have access to the AAR ( but not prefab modules yet);
+   again this is not Prefab specific either
+1. Enable Prefab in the gradle scripts to access prefab module(s) inside AAR
+1. Now you could use the Prefab modules in your app (in the source files and C/C++ build system)
```
make sense, updated, thanks.

DanAlbert replied:
This wasn't fixed.

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,80 @@
+
+* prefab-publishing is for how to create Prefab dependency into an AAR
+* prefab-dependency is for how to use Prefab AARs hosted in Google Maven Center
+* curl-ssl is for how to use 2 very important AARs (curl and ssl )
```
done

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,80 @@
+1. Now you could use the Prefab modules in your app (in the source files and C/C++ build system)
+
+Note again that the step 2 and 3 above are not C/C++ or prefab specific, please refer to
+[the formal AAR documentations] for details. The rest of the steps are covered in the Prefab samples here:
```
done (whatever appears on Developer.Android.Com is formal documentation though)

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,80 @@
+Note again that the step 2 and 3 above are not C/C++ or prefab specific, please refer to
+[the formal AAR documentations] for details. The rest of the steps are covered in the Prefab samples here:
+
+* prefab-publishing is for how to create Prefab dependency into an AAR
```
done

DanAlbert replied:
not fixed

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,80 @@
+[the formal AAR documentations] for details. The rest of the steps are covered in the Prefab samples here:
+
+* prefab-publishing is for how to create Prefab dependency into an AAR
+* prefab-dependency is for how to use Prefab AARs hosted in Google Maven Center
```
let's use GMave here, updated and added link too.

ggfan commented:
yeah, addressed code review items ( a couple of discussions there too ).  PTAL, thank you!

DanAlbert reviewed (changes requested):
There are still a ton of not-yet-fixed comments from my last review.

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,80 @@
+Prefab Availability
+-------------------
+
+Prefab has been added into Android Gradle Plugin since version 4:
```
done

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,80 @@
+-------------------
+
+Prefab has been added into Android Gradle Plugin since version 4:
+* Using Prefab AARs: AGP 4.0+ (with some update in 4.1+)
```
the enabling part: it was inside gradle.properties, now inside build.gradle. that is what it means.

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,80 @@
+
+Prefab has been added into Android Gradle Plugin since version 4:
+* Using Prefab AARs: AGP 4.0+ (with some update in 4.1+)
+* Generating Prefab AARs: AGP 4.1+
```
done

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,80 @@
+* Using Prefab AARs: AGP 4.0+ (with some update in 4.1+)
+* Generating Prefab AARs: AGP 4.1+
+
+The prefab AARs used by the samples here are hosted at [Google Maven], but you could
```
done

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,81 @@
+Pre-requisites
+--------------
+
+* Android Gradle Plugin 4.0+/4.1+
```
for generation part, it must be 4.1+.  yeah will leave it to the generation sample

ggfan commented on prefab/curl-ssl/README.md:
done

ggfan commented on prefab/curl-ssl/README.md:
works, changed. thx

ggfan commented on prefab/curl-ssl/README.md:
done

ggfan commented on prefab/curl-ssl/settings.gradle:
```diff
@@ -1,2 +1,2 @@
-rootProject.name='curl-ssl'
+rootProject.name='prefab-using'
```
Oops, this is good one: fixed. thanks!

ggfan commented on prefab/prefab-publishing/README.md:
```diff
@@ -1,27 +1,55 @@
-
-To export native libraries in an AAR with Android Gradle Plugin 4.1+, make the
-following changes to your Android Library project's [build.gradle]:
+This sample shows how to export native libraries to an AAR
+using the [Prefab] format. Generally speaking, to export native libraries in an AAR 
+with Android Gradle Plugin 4.1+, make the following changes to your Android Library project's [build.gradle]:
```
As a user, I think it added more details and "could follow and get it done".

ggfan replied:
updated

ggfan commented on prefab/prefab-publishing/README.md:
````diff
@@ -1,27 +1,55 @@
 
 * Enable the `prefabPublishing` build feature
+```
+    android.buildFeatures.prefabPublising true
````
https://github.com/android/ndk-samples/blob/prefab-update/prefab/prefab-publishing/README.md
rendering seems ok to me.

ggfan commented:
went through the list again, did fix them, may you have a look?  thanks

DanAlbert commented on prefab/prefab-publishing/README.md:
````diff
@@ -1,27 +1,52 @@
+```
+This sample builds a trivial library called `mylibrary` (libmylibrary.so), and packs it into 
+mylibrary.aar. From here you could distribute the AAR via Maven or however you distribute AARs.
+To learn how to use the generated dependency AAR, refer to [the prefab-dependency sample] in this repo;
````
To learn how to use the genreated AAR

ggfan replied:
done

DanAlbert commented on prefab/prefab-publishing/mylibrary/build.gradle:
```diff
@@ -39,14 +39,27 @@ android {
         }
     }
 
+    // Enable creation of Prefab packages and integration into AAR.
```
Enable generation of Prefab packages and include them in the library's AAR.

ggfan replied:
done

DanAlbert commented on prefab/prefab-publishing/mylibrary/build.gradle:
```diff
@@ -39,14 +39,27 @@ android {
         prefabPublishing true
     }
 
+    // Configure prefab to include the "mylibrary" module from the native build system,
+    // and export the headers in src/main/cpp/include.
```
Include the "mylibrary" module from the native build system in the AAR, and export the headers in src/main/cpp/include to its consumers.

ggfan replied:
done

ggfan commented on prefab/curl-ssl/README.md:
done.

ggfan commented on prefab/curl-ssl/README.md:
```diff
@@ -1,34 +1,38 @@
-The curl-ssl sample shows how to import native dependencies from Maven. This
-sample uses [curl], [OpenSSL], and [JsonCpp] to display a list of the most
-recent 10 reviews submitted to AOSP's code review system.
+This sample shows how to import [curl] and [OpenSSL] C/C++ dependencies from Google Maven to your Android project
+to display a list of the most recent 10 reviews submitted to AOSP's code review system.
```
updated.

ggfan commented on prefab/curl-ssl/README.md:
```diff
@@ -1,34 +1,38 @@
-# curl-ssl
+curl-ssl
+========
```
ok, thanks

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,87 @@
+* Using Prefab AARs requires AGP 4.0+ (with some update in 4.1+)
+* Generating AARs that export native dependencies requires AGP 4.1+
+
+The prefab AARs used by the samples here are hosted at [Google Maven], but you can
```
updated

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,87 @@
+-------------------
+
+Support for Prefab packages has been added available Android Gradle Plugin since version 4.0:
+* Using Prefab AARs requires AGP 4.0+ (with some update in 4.1+)
```
updated. ( prefab enabling has changed, right? from gradle.properties --> build.gradle, you have a warning for that in 4.1 )

DanAlbert replied:
You weren't saying that though. Fine to document if you think it helps, but that's not what this said.

I'm also fine with not mentioning it here as you've done. We don't need to copy all the docs here.

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,87 @@
+Prefab Availability
+-------------------
+
+Support for Prefab packages has been added available Android Gradle Plugin since version 4.0:
```
thanks: this was missed.

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,87 @@
+With that in mind, the samples here collectively demonstrate prefab usage with:
+
+* prefab-publishing: shows how to create Prefab dependency into an AAR
+* prefab-dependency: shows how to import native dependencies form [GMaven]
```
yeah, thanks!

ggfan commented on prefab/README.md:
```diff
@@ -0,0 +1,87 @@
+
+With that in mind, the samples here collectively demonstrate prefab usage with:
+
+* prefab-publishing: shows how to create Prefab dependency into an AAR
```
done

ggfan commented:
Addressed all reviews, one more look please

DanAlbert commented on prefab/curl-ssl/README.md:
HTTPS

ggfan replied:
Oops, it is really a typo.  fixed in the next upload

DanAlbert commented on prefab/README.md:
```diff
@@ -0,0 +1,87 @@
+[Prefab format] and packed inside AARs. Conceptually, Prefab's flow in Android Gradle Plugin,
+at high level, could be viewed as:
+
+1. Use AGP to pack C/C++ libraries (including the header files), called modules, into AAR
+1. Distribute AARs with maven center or other mechanisms, but this is not Prefab specific
+1. Add AAR dependency to your application, now you have access to the AAR (but not prefab modules yet);
+   again this is not Prefab specific either
+1. Enable Prefab in the gradle scripts to access prefab module(s) inside AAR
+1. Use Prefab modules in your native build system such as CMake or ndk-build
```
Not fixed yet.

ggfan commented:
one more review please, thanks

DanAlbert commented on prefab/README.md:
```diff
@@ -0,0 +1,90 @@
+Prefab Samples
+==============
+
+The samples here are to demontrate how to use [the Prefab feature] in Android Gradle Plugin.
+Generally speaking, Prefab is a mechanism to make C/C++ libraries available to the gradle's
+C/C++ build system, with the requirement that the C/C++ libraries are organized in the
+[Prefab format] and packed inside AARs. In order to use the C/C++ dependency AAR, as AAR authors,
+you need to do the following:
+
+1. Use AGP to pack C/C++ libraries (including the header files), called modules, into AAR
+1. Distribute AARs with [Google Maven] or other mechanisms, but this is not Prefab specific
+1. Add AAR dependency to your application, now you have access to the AAR (but not prefab modules yet);
+   again this is not Prefab specific either
+1. Enable Prefab in the gradle scripts to access prefab module(s) inside AAR
+1. Use Prefab modules in your native build system such as CMake or ndk-build
+
+However as AAR users, you just need to do step 3 through step 5 above, i.e:
+1. Add AAR dependency to your application, now you have access to the AAR (but not prefab modules yet);
+   again this is not Prefab specific either
+1. Enable Prefab in the gradle scripts to access prefab module(s) inside AAR
+1. Use Prefab modules in your native build system such as CMake or ndk-build
+
+Note that the step 2 and 3 above are not C/C++ or prefab specific, please refer to
+[the AAR documentation] for details.
```
The samples in this directory demonstrate how to create and consume C/C++
dependencies with the Android Gradle Plugin. Dependencies are packaged within
AARs using the [Prefab] package format. The Android Gradle Plugin natively
supports producing and consuming these, so you do not need to worry about the
details of the packaging format.

The individual samples will explain in more detail, but the typical workflow for
a consumer of a native dependency is:

1. Enable the `prefab` build feature in your `build.gradle` file.
2. Add the dependency to the `dependencies` block of your build.gradle file.
3. Import the package in your `CMakeLists.txt` or `Android.mk` file.
4. Link the dependencies to your libraries.

Linking the dependency to your library will automatically make the headers
available, link the required libraries, and include the dependencies you use in
your APK or AAR.

To produce an AAR that exposes C/C++ APIs:

1. Enable the `prefabPublishing` build feature in your `build.gradle` file.
2. Use `android.prefab` in your `build.gradle` file to declare the names of the
   libraries you wish to export and the headers that define their interface.

To test your packages, follow the steps for package consumers. Note that until
https://issuetracker.google.com/120166563 is fixed it is not possible to depend
on the native components of an AAR generated by another module of the same
project when using Android Studio. Instead, test in a separate project. You can
import the AAR into your test project either [directly] or by publishing it to a
Maven repository.

[directly]: https://developer.android.com/studio/projects/android-library#AddDependency
[Prefab]: https://google.github.io/prefab

ggfan replied:
done

DanAlbert commented on prefab/curl-ssl/README.md:
```diff
@@ -1,34 +1,38 @@
-The curl-ssl sample shows how to import native dependencies from Maven. This
-sample uses [curl], [OpenSSL], and [JsonCpp] to display a list of the most
-recent 10 reviews submitted to AOSP's code review system.
+This sample shows how to import [curl] and [OpenSSL] C/C++ dependencies from Google Maven
```
This sample shows how to import [curl] and [OpenSSL] from Google Maven...

ggfan replied:
done

DanAlbert commented on prefab/prefab-publishing/.clang-format:
```diff
@@ -1 +0,0 @@
-BasedOnStyle: Google
```
Why? This is useful.

ggfan replied:
no, it is command line parameter, right.

DanAlbert replied:
We have command line users. It's also usable by Android Studio: https://www.jetbrains.com/help/clion/clangformat-as-alternative-formatter.html#config-files

ggfan replied:
out of ~200 samples, no one checks in this file except for ios samples. Would be good to keep the tradition aligned with the rest of the Android samples.

DanAlbert replied:
Yeah, it would be good to add it to all of the other samples. Could move this to the top level instead if you prefer.

DanAlbert commented on prefab/README.md:
```diff
@@ -0,0 +1,98 @@
+The individual samples will explain in more detail, but the typical workflow for
+a consumer of a native dependency is:
+
+1. Enable the prefab build feature in your build.gradle file.
+1. Add the dependency to the dependencies block of your build.gradle file.
+1. Import the package in your CMakeLists.txt or Android.mk file.
+1. Link the dependencies to your libraries.
+
+Linking the dependency to your library will automatically make the headers
+available, link the required libraries, and include the dependencies you use in
+your APK or AAR.
+
+To produce an AAR that exposes C/C++ APIs:
+
+1. Enable the prefabPublishing build feature in your build.gradle file.
+1. Use android.prefab in your build.gradle file to declare the names of the
+libraries you wish to export and the headers that define their interface.
```
Looks like the code formatting that was in the text I posted got dropped.

ggfan replied:
added back some code format, modified the "Linking the .... " sentence clarify what it means ( it is hard one when suddenly dropping another "your APK or AAR" when this is mainly talking about how to use AAR, changed that one to "application", application in this sense could be APK or AAR, but this AAR has NOTHING to do with native dependency AAR)

DanAlbert commented on prefab/README.md:
```diff
@@ -0,0 +1,98 @@
+import the AAR into your test project either [directly](https://developer.android.com/studio/projects/android-library#AddDependency)
+or by publishing it to a Maven repository.
+
+With that in mind, the samples here collectively demonstrate prefab usage with:
```
drop "with" at the end. It doesn't fit the phrasing of the bullet points.

ggfan replied:
done

DanAlbert commented on prefab/README.md:
```diff
@@ -1,27 +1,27 @@
-available, link the required libraries, and include the dependencies you use in
-your APK or AAR.
+available, link the required libraries, and pack the dependencies you use into
+your application.
```
or library (or go back to the previous "APK or AAR" wording)

ggfan replied:
changed

ggfan commented:
for .clang-format, will report it out to the codelab council, if adopted, would update together with other samples.  let's move on for this one and address it later, thank you.

DanAlbert commented on prefab/README.md:
```diff
@@ -1,27 +1,27 @@
 Linking the dependency to your library will automatically make the headers
-available, link the required libraries, and include the dependencies you use in
-your APK or AAR.
+available, link the required libraries, and pack the dependencies you use into
```
previous wording "include the dependencies you use in" was better

ggfan replied:
changed

DanAlbert commented:
> for .clang-format, will report it out to the codelab council, if adopted, would update together with other samples. let's move on for this one and address it later, thank you.

If you want to move on from it drop that part of the change. It's unrelated to the PR.

ggfan commented:
.clang-format moved to the top of the directory `prefab/` for now. thanks

salahalsheharicloud reviewed (commented):
[iMessage-Business-Chat.zip](https://github.com/android/ndk-samples/files/5759483/iMessage-Business-Chat.zip)
#
